### PR TITLE
Issue 5770 - RFE - Extend Password Adminstrators to allow skipping password info updates

### PR DESCRIPTION
**Pull request by mreynolds389**
Description:

Add new config setting to state that password admin updates should not update entry's password state attributes.

relates: https://github.com/389ds/389-ds-base/issues/5770

### CODE REVIEW

**Comment by progier389** [comment]
Looks good but I would expect that "dsconf instance pwpolicy set/get" get changed too ...

**Comment by mreynolds389** [comment]
> Looks good but I would expect that "dsconf instance pwpolicy set/get" get changed too ...

Yeah and the UI...  Working on it...

File: ldap/servers/slapd/modify.c [hunk 1]
```diff
@@ -1207,7 +1207,7 @@ op_shared_allow_pw_change(Slapi_PBlock *pb, LDAPMod *mod, char **old_pw, Slapi_M
         slapi_pblock_set(pb, SLAPI_BACKEND, slapi_be_select(&sdn));
 
         /* Check if ACIs allow password to be changed */
-        if (!pw_is_pwp_admin(pb, pwpolicy) && (res = slapi_acl_check_mods(pb, e, mods, &errtxt)) != LDAP_SUCCESS) {
+        if (!pw_is_pwp_admin(pb, pwpolicy, PR_FALSE) && (res = slapi_acl_check_mods(pb, e, mods, &errtxt)) != LDAP_SUCCESS) {
```
Comment (tbordaz) [inline ldap/servers/slapd/modify.c]: Shouldn't it be PWP_INCLUDE_ROOTDN rather than PR_FALSE ?

File: ldap/servers/slapd/pw.h [hunk 2]
```diff
@@ -41,7 +41,10 @@ int check_pw_duration_value(const char *attr_name, char *value, long minval, lon
 int check_pw_resetfailurecount_value(const char *attr_name, char *value, long minval, long maxval, char *errorbuf, size_t ebuflen);
 int check_pw_storagescheme_value(const char *attr_name, char *value, long minval, long maxval, char *errorbuf, size_t ebuflen);
 
-int pw_is_pwp_admin(Slapi_PBlock *pb, struct passwordpolicyarray *pwp);
+int pw_is_pwp_admin(Slapi_PBlock *pb, struct passwordpolicyarray *pwp, int ignore_rootdn);
+#define PWP_INCLUDE_ROOTDN 0
+#define PWP_IGNORE_ROOTDN 1
+
```
Comment (tbordaz) [inline ldap/servers/slapd/pw.h:47]: My understanding is that this flag is used to check if the requestor is root || pwp_admin (INCLUDE_ROOTDN) or pwp_admin only.
I have no strong opinion but I would prefer something like PWP_ADMIN_ONLY and PWP_ADMIN_OR_ROOT

Comment (mreynolds389) [reply ldap/servers/slapd/pw.h:47]: Ok, I'll rewrite it

File: ldap/servers/slapd/modify.c (hunk 1 above)
Comment (mreynolds389) [reply ldap/servers/slapd/modify.c]: Nice catch, it will be fixed

File: ldap/servers/slapd/pw.c [hunk 3]
```diff
@@ -1903,7 +1906,7 @@ pw_is_pwp_admin(Slapi_PBlock *pb, passwdPolicy *pwp)
     slapi_pblock_get(pb, SLAPI_REQUESTOR_ISROOT, &is_requestor_root);
 
     /* first check if it's root */
-    if (is_requestor_root) {
+    if (is_requestor_root && !ignore_rootdn) {
```
Comment (tbordaz) [inline ldap/servers/slapd/pw.c]: It could also be something like 
if (is_requestor_root && (ignore_rootdn == PWP_ADMIN_OR_ROOTDN)) {

**Review by tbordaz — APPROVED** [review]
The server code looks good to me. (minor comments)

File: src/lib389/lib389/cli_conf/pwpolicy.py [hunk 4]
```diff
@@ -237,6 +237,9 @@ def create_parser(subparsers):
     set_parser.add_argument('--pwdhistory', help="To enable password history set this to \"on\", otherwise \"off\"")
     set_parser.add_argument('--pwdhistorycount', help="The number of passwords to keep in history")
     set_parser.add_argument('--pwdadmin', help="The DN of an entry or a group of account that can bypass password policy constraints")
+    set_parser.add_argument('--pwdadminskipupdates',
+                            help="Set to \"on\" if the Password Admin's password update should not trigger updates to the passowrd state attributes (passwordExpirationtime, passwordHistory, etc).")
```
Comment (droideck) [inline src/lib389/lib389/cli_conf/pwpolicy.py]: `passowrd` - typo.

But also, you've added the setting to local pwpolicy in CLI, but the UI doesn't have it in local pwpolicy...
So I'm a bit confused if we want it in local pwpolicy or not?

**Review by droideck — CHANGES_REQUESTED** [review]

File: src/lib389/lib389/cli_conf/pwpolicy.py (hunk 4 above)
Comment (mreynolds389) [reply src/lib389/lib389/cli_conf/pwpolicy.py]: ok, the UI had half implemented it in the local policy.  It is all fixed and consistent now

**Review by droideck — APPROVED** [review]
Great! LGTM!

**Pull request merged**